writer.go

```diff
@@ -23,11 +23,15 @@ func (mw *messageWriter) writeLine(l []byte) (int, error) {
 }
 
 func (mw *messageWriter) Write(p []byte) (int, error) {
+	// N will count the number of bytes *from p* that were written. Since we'll
+	// scan (and increment N for) all the bytes already in the buffer before
+	// the write and those in p, N should start at minus mw.buf.Len().
+	N := -mw.buf.Len()
```

Comment:
I'm worried about returning negative numbers if we hit an error. For instance, if `mw.buf.Len()` is originally 10, and we want to write 5 bytes more, the underlying `io.Writer` may fail at byte 2 out of the total 15 bytes, and then we return -8.

Maybe we could handle this outside of the loop, by replacing `return N, err` by `break`?

```go
var (
  err error
  N int
)
for {
  // …
}

N -= len(b) - len(p)
if N < 0 {
  N = 0
}
return N, err
```